python/ray/train/base_trainer.py

```diff
@@ -400,18 +475,32 @@ def _generate_trainable_cls(self) -> Type["Trainable"]:
         trainer_cls = self.__class__
         scaling_config = self.scaling_config
 
-        def train_func(config, checkpoint_dir=None):
+        def train_func(config):
             # config already contains merged values.
             # Instantiate new Trainer in Trainable.
             trainer = trainer_cls(**config)
 
-            if checkpoint_dir:
-                trainer.resume_from_checkpoint = Checkpoint.from_directory(
-                    checkpoint_dir
+            # Get the checkpoint from the Tune session, and use it to initialize
+            # the restored trainer
+            checkpoint = session.get_checkpoint()
+            loaded_preprocessor = False
+            if checkpoint:
+                trainer.resume_from_checkpoint = checkpoint
+                # Always use the checkpointed preprocessor when retrying on errors
+                # (configured by FailureConfig(max_failures)).
+                # On a restored the full experiment via Trainer.restore, only load the
+                # checkpointed preprocessor if the preprocessor is not re-specified.
+                should_load_from_checkpoint = (
+                    not trainer._restore_path or not trainer.preprocessor
```

Comment:
name this ``should_load_preprocessors_from_checkpoint``?
also the logic seems weird, we will do that if _restore_path is None, regardless of trainer.preprocessor state?

Reply:
Yep you're right, should just be `should_load_preprocessor_from_checkpoint = not trainer.preprocessor`.